server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.profile.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link CollectorManager} that takes another CollectorManager as input and wraps all Collectors generated by it
+ * in an {@link InternalProfileCollector}. It delegates all the profiling to the generated collectors via {@link #getCollectorTree()}
+ * and joins them up when its {@link #reduce} method is called. The profile result can
+ */
+public final class ProfileCollectorManager implements CollectorManager<InternalProfileCollector, Void> {
+
+    private final CollectorManager<Collector, Void> collectorManager;
+    private CollectorResult collectorTree;
+
+    private ProfileCollectorManager(CollectorManager<Collector, Void> collectorManager) {
+        this.collectorManager = collectorManager;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <C extends Collector> ProfileCollectorManager create(CollectorManager<C, ?> collectorManager) {
+        return new ProfileCollectorManager((CollectorManager<Collector, Void>) collectorManager);
+    }
+
+    @Override
+    public InternalProfileCollector newCollector() throws IOException {
+        return new InternalProfileCollector(collectorManager.newCollector(), CollectorResult.REASON_SEARCH_TOP_HITS);
```

Comment:
can you make the reason configurable instead of hardcoded?

Reply:
Also, in #96622 I added support for setting the child collectors as well. Could you double check if that makes sense and integrate it?

Reply:
I added the "reason" to the constructor. As for the child collectors, I can see where this will probably be needed later in your PR but I would prefer to add it there, so the motivation is clear.